modo/api.py

```diff
@@ -321,3 +323,33 @@ def enrich_metadata(self):
                     self.update_element(inst_names[ele.name], ele)
                 else:
                     continue
+
+    def stream_cram(
+        self, cram_name: str, region: str = None, output: str = None
+    ):
+        """Slices and streams the requested CRAM file, both local and remote,
+        and either outputs a data stream or writes data to local file"""
+
+        # check requested CRAM exists in MODO
+        path = ""
+        filepaths = list(self.list_files())
+        for filepath in filepaths:
+            if cram_name == str(filepath).split("/")[-1]:
+                path = filepath
+                break
+        if path == "":
+            raise ValueError(f"{cram_name} not fount in {self.path}.")
+
+        if self.s3_endpoint:
+            # http://domain/s3 + bucket/modo/file.cram --> http://domain/htsget/reads/modo/file.cram
+            url = (
+                re.sub(r"s3$", "", self.s3_endpoint)
```

Comment:
The user should have the option to specify a custom s3/htsget url. If not specified, we'll infer via substitution (with warning msg maybe). Would that work? @AssafSternberg

Reply:
Yes, that should work.

instead of,
modo=MODO(path= 'modo-demo/ex', s3_endpoint = 'http://localhost/s3')

we can use,
modo=MODO(path= 'modo-demo/ex', s3_endpoint = 'http://localhost/s3', htsget_endpoint = 'http://htsget_domain/')

Reply:
Looks good yes, the default value can be None, and the docstring explain that this means falling back to same domain

Reply:
BTW, just noticed that we don't actually initialize 'self.s3_endpoint', rather 's3_endpoint' is just used as a variable. I think we should add 'self.s3_endpoint', and as discussed with Cyril this morning, also a 'self.htsget_endpoint', which can be then used for translating between the s3 path and the htsget path, without needing the users to enter them, and averting problems if they enter wrong or invalid endpoints.

Reply:
I added self.s3_endpoint and self.htsget_endpoint to MODO. If not given, self.htsget_endpoint is assumed to have the same domain as s3_endpoint, or None if s3_endpoint is not given as well (see commit - https://github.com/sdsc-ordes/modo-api/pull/31/commits/aa7fc9712fa89a6a5ab3d23457fb552a5ab1c0a5)

`
class MODO:
    """Multi-Omics Digital Object
    A digital archive containing several multi-omics data and records.
    The archive contains:
    * A zarr file, array-based data and metadata pointing to arrays and data files
    * CRAM files, with genomic-alignments data
    Examples
    --------
    >>> demo = MODO("data/ex")
    # List identifiers of samples in the archive
    >>> demo.list_samples()
    ['/sample/sample1']
    # List files in the archive
    >>> files = sorted([file.name for file in demo.list_files()])
    >>> assert 'demo1.cram' in files
    >>> assert 'reference1.fa' in files
    """
    def __init__(
        self,
        path: Union[Path, str],
        s3_endpoint: Optional[str] = None,
        htsget_endpoint: Optional[str] = None,
        id: Optional[str] = None,
        name: Optional[str] = None,
        description: Optional[str] = None,
        creation_date: date = date.today(),
        last_update_date: date = date.today(),
        has_assay: List = [],
        source_uri: Optional[str] = None,
    ):
        self.s3_endpoint = s3_endpoint
        if s3_endpoint and not htsget_endpoint:
            htsget_endpoint = re.sub(r"s3$", "htsget", s3_endpoint)
        self.htsget_endpoint = htsget_endpoint`